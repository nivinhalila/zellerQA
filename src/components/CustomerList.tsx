import { useQuery } from "@apollo/client";
import React from "react";
import styled from "styled-components";
import StringOperators from '../utils/StringOperators'

import CustomerListItem from "./CustomerListItem";
import Section from "./Section";
import Text from "./Text";
import Title from "./Title";

import { ListZellerCustomersData } from "../types";
import { LIST_ZELLER_CUSTOMERS } from "../graphql/queries";

const ErrorText = styled(Text)`
  color: red;
  font-size: 1.2em;
`;

const LoadingText = styled(Text)`
  color: #adadac;
  font-size: 1.2em;
`;

interface CustomerListProps {
  role: string;
}

export function CustomerList(props: CustomerListProps) {
  const { role } = props;

  // NOTE: the apollo client will return cached local results following the
  // initial successful fetch, to ensure the data is up to date you are able to
  // refetch this on command or poll the backend, I've left this out to keep
  // this simple for now.
  const { loading, error, data } = useQuery<ListZellerCustomersData>(
    LIST_ZELLER_CUSTOMERS
  );

  // NOTE: the graphapi call returns all customers regardless of role type, to
  // remove the filtering below and improve performance/efficiency I would
  // update the api call to return all users of a particular role type.
  const items = data?.listZellerCustomers?.items || [];
  const customersByRole = items.filter((item: any) => item.role === role);
  const title = `${role} Users`;

  if (loading) {
    return (
      <Section>
        <Title>{title}</Title>
        <LoadingText>Loading users...</LoadingText>
      </Section>
    );
  }

  if (error) {
    return (
      <Section>
        <Title>{title}</Title>
        <ErrorText>
          We were unable to complete your request at this time. Please try again
          or contact support.
        </ErrorText>
      </Section>
    );
  }

  return (
    <Section>
      <Title>{title}</Title>
      {customersByRole.map((customer) => (
        <CustomerListItem
          key={customer.id}
          name={customer.name}
          role={StringOperators.toTitleCase(customer.role)}
        />
      ))}
    </Section>
  );
}

export default CustomerList;
