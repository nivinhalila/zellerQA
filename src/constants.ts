import { ZellerAdmin, ZellerManager } from "./types";

export const ZELLER_ADMIN: ZellerAdmin = "ADMIN";
export const ZELLER_MANAGER: ZellerManager = "MANAGER";
