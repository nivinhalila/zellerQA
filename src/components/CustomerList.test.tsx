import "jest-styled-components";
import { MockedProvider } from "@apollo/client/testing";
import { render, waitFor } from "@testing-library/react";

import CustomerList from "./CustomerList";
import { ZELLER_ADMIN } from "../constants";
import { LIST_ZELLER_CUSTOMERS } from "../graphql/queries";

test("renders role title header", () => {
  expect.assertions(1);

  const role = "Superadmin";

  const { getByText } = render(
    <MockedProvider addTypename={false}>
      <CustomerList role={role} />
    </MockedProvider>
  );

  expect(getByText("Superadmin Users")).toBeInTheDocument();
});

test("renders customer list component when request loading", () => {
  expect.assertions(2);

  const role = ZELLER_ADMIN;

  const { container, getByText } = render(
    <MockedProvider addTypename={false}>
      <CustomerList role={role} />
    </MockedProvider>
  );

  // loading text
  expect(getByText("Loading users...")).toBeInTheDocument();

  expect(container.firstChild).toMatchSnapshot();
});

test("renders customer list component when request errors", async () => {
  expect.assertions(2);

  const listCustomersMock = {
    request: {
      query: LIST_ZELLER_CUSTOMERS,
    },
    error: new Error("Uh oh!"),
  };

  const mocks = [listCustomersMock];
  const role = ZELLER_ADMIN;

  const { container, findByText } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <CustomerList role={role} />
    </MockedProvider>
  );

  const errorText = await findByText(
    "We were unable to complete your request at this time. Please try again or contact support."
  );

  expect(errorText).toBeInTheDocument();

  expect(container.firstChild).toMatchSnapshot();
});

test("renders customer list when request successful", async () => {
  expect.assertions(3);

  const listCustomersMock = {
    request: {
      query: LIST_ZELLER_CUSTOMERS,
    },
    result: {
      data: {
        listZellerCustomers: {
          items: [
            {
              email: "customer@zeller.com",
              id: "customer1",
              name: "Jonathan Warykowski",
              role: ZELLER_ADMIN,
            },
          ],
        },
      },
    },
  };

  const mocks = [listCustomersMock];

  const { container, queryByText } = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <CustomerList role={ZELLER_ADMIN} />
    </MockedProvider>
  );

  await waitFor(() => {
    expect(queryByText("Loading users...")).not.toBeInTheDocument();
  });

  expect(container.firstChild).toMatchSnapshot();
});
