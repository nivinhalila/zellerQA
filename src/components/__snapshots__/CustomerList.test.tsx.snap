// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders customer list component when request errors 1`] = `
.c2 {
  color: #212322;
  font-size: 1.5em;
}

.c0 {
  background: #ffffff;
  margin-bottom: 2.5rem;
}

.c1 {
  color: #212322;
  font-size: 2em;
  font-weight: bold;
}

.c3 {
  color: red;
  font-size: 1.2em;
}

<section
  class="c0"
>
  <h1
    class="c1"
  >
    ADMIN Users
  </h1>
  <p
    class="c2 c3"
  >
    We were unable to complete your request at this time. Please try again or contact support.
  </p>
</section>
`;

exports[`renders customer list component when request loading 1`] = `
.c2 {
  color: #212322;
  font-size: 1.5em;
}

.c0 {
  background: #ffffff;
  margin-bottom: 2.5rem;
}

.c1 {
  color: #212322;
  font-size: 2em;
  font-weight: bold;
}

.c3 {
  color: #adadac;
  font-size: 1.2em;
}

<section
  class="c0"
>
  <h1
    class="c1"
  >
    ADMIN Users
  </h1>
  <p
    class="c2 c3"
  >
    Loading users...
  </p>
</section>
`;

exports[`renders customer list when request successful 1`] = `
.c4 {
  color: #212322;
  font-size: 1.5em;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #e8f2fb;
  border-radius: 0.5rem;
  margin: 0 1.5rem 0 0;
  width: 3.5rem;
  height: 3.5rem;
}

.c5 {
  color: #005cc8;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: 0;
  text-align: center;
}

.c7 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: 0;
}

.c8 {
  color: #787775;
  font-size: 1.2em;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin: 0;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: "row";
  -ms-flex-direction: "row";
  flex-direction: "row";
  margin-bottom: 1.5rem;
}

.c2:last-child {
  margin-bottom: 0;
}

.c0 {
  background: #ffffff;
  margin-bottom: 2.5rem;
}

.c1 {
  color: #212322;
  font-size: 2em;
  font-weight: bold;
}

<section
  class="c0"
>
  <h1
    class="c1"
  >
    ADMIN Users
  </h1>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <p
        class="c4 c5"
      >
        J
      </p>
    </div>
    <div
      class="c6"
    >
      <p
        class="c4 c7"
      >
        Jonathan Warykowski
      </p>
      <p
        class="c4 c8"
      >
        Admin
      </p>
    </div>
  </div>
</section>
`;
